src/irmin-pack/unix/ext.ml

```diff
@@ -30,16 +30,11 @@ module Maker (Config : Conf.S) = struct
     end
 
     module H = Schema.Hash
-    module Index = Pack_index.Make (H)
     module Io = Io.Unix
+    module Index = Pack_index.Make (H)
+    module Mapping_file = Mapping_file.Make (Io)
```

Comment:
What's the motivation for moving mapping file out of file manager when the other files (control, aof, suffix) are going the other direction?

Reply:
The idea was to move away from File Manager, all the modules that are forced to use `Io.Unix`, while removing the other functors (aof & co), as they are not dependent on a specific implementation of Io.

The ideal situation would be to have `Mapping_file` & `Index` independent of `Io.Unix`, therefore only functorizing FM with an Io implementation, but if it is possible with `Mapping_file` (with some work), it is way more costly and might be hugely bug inducing with `Index` (We discussed it a lot, Ngoguey42 and me)

Reply:
Thanks for the extra information.

It sounds like this PR came out of a larger discussion/planning, but if an idea is to functorize `File_manager` with just `Io.S`, it is possible to keep `Mapping_file` in `File_manager` by changing `Mapping_file.Make` to not force `Io.Unix` since this isn't strictly necessary and works fine if it is just `Io.S` (tested locally just now). It is, however, true that `Mapping_file` very much depends on `Unix` in its implementation, which is perhaps the delineation being made here.

All that to say, it is not immediately clear to me how best to parameterize `File_manager` inside the scope of `irmin-pack.unix`. Since it is already in a "unix" scope, it would be reasonable to me to parameterize with just one parameter (either `Index` or `H` to make `Index`) and then use `Io.Unix` and `Make` for every module it needs.

@Ngoguey42 as the original author of this suite of modules, do you have some ideas of how you'd like to see them evolve that you could tl;dr?

(None of this discussion blocks this PR to me -- just useful to have the conversation :smile:)

Reply:
The utopic goal is to one day split irmin-pack.unix into 2 in order to allow an irmin-pack.mirage.

There are 2 ways forward:
1. completely drop that dream which would simplify the functor stack here or
2. keep the dream alive by having some work already done in the code.

I think we shouldn't sweat too much about these diffs since they only solve some local inconsistencies without tackling the deeper problem (i.e. implementing for mirage or dropping the idea)

Reply:
> The utopic goal is to one day split irmin-pack.unix into 2 in order to allow an irmin-pack.mirage.

But I don't think the current IO signature will ever be efficiently ported to mirage, [io.mli](https://github.com/mirage/irmin/blob/main/src/irmin-pack/unix/io_intf.ml) is a file-based API, which makes a lots of assumption on the POSIX semantics. To have `irmin-pack.mirage`, we need to have a block-device API. 

So it's no an utopic goal, but this requires investigation and prototyping work first. The approach I would take is to find the right abstractions and move them to `irmin`'s core. And create a specialised block-device API for mirage, while keeping an efficient file-based API for UNIX. So, unless you have a better plan, I reckon you should aim to simplify the Unix code first and make it more maintainable.

Reply:
@Ngoguey42 I had a hunch your mind might have been towards mirage.

@samoht your points about file-based vs block-based are useful for thinking about posix-type target vs mirage. Do you know of any mirage-based projects that have more advanced usages that could be inspiration for a multi-file backend like irmin-pack? 

The current functor stack is not a huge maintenance burden but could be simplified some if we assume unix/posix/files. We can address that if we desire in future PRs.

Reply:
There are various ways to emulate files for mirage (for instance, https://github.com/yomimono/chamelon), but the simplest way to emulate files is not to have them at all. The current basic blocks of Irmin (like append-only stores) are very well-suited for block devices when you don't need a very complicated structure (ie. _not_ POSIX). For mirage, it's much easier to have everything static: allocate a small set of initial block devices (that are just contiguous sets of bytes) and build the minimal structure that you need on top of this to support the AO and RW abstractions.

I also think we should drastically simplify the semantics for mirage: we don't have to worry about concurrent read access at the block device level. For concurrent accesses, force users to use `irmin-server` and mediate concurrency via higher-level APIs. You also don't have to worry too much about persistency: there is no kernel page cache, so every write is directly propagated to the (virtual) block device - as the result, there is no need for `fsync` - although you might need to have write barriers primitives in some circumstances (there was a proposal here: https://github.com/mirage/wodan/issues/14).

Reply:
To complete this: the closer thing to a filesystem is https://github.com/mirage/mirage-kv/blob/main/src/mirage_kv.mli (that's implemented by various backends - chameleon is one of them) but again I don't think that's the right approach to port irmin-pack to mirage.

Reply:
> I reckon you should aim to simplify the Unix code first and make it more maintainable.

I agree